### ISSUE TITLE
PHP Warning:  fclose(): supplied resource is not a valid stream resource

### ISSUE THREAD

**Issue by IT-Experte** [mongodb/mongo-php-library#311]
PHP 7.0.8
mongo-php-library 1.1.0 (final)
---
PHP Warning:  fclose(): supplied resource is not a valid stream resource in .../src/GridFS/ReadableStream.php on line 78
---

This is my php code. It reads images from a MongoDB collection and display each file.

```php
<?php
// include Composer (mongodb) extension
require '/src/composer/vendor/autoload.php';

class GridFS {

        private $client;

        public function __construct() {
                $this->client = new MongoDB\Client("mongodb://localhost:27017");
        }

        public function Image($id) {

                $db = $this->client->selectDatabase("myfiles");
                $bucket = $db->selectGridFSBucket();

                $stream = $bucket->openDownloadStream($id);
                header("Content-type: image/jpg;");
                echo stream_get_contents($stream);
                exit;
        }

        public function FindFiles($limit) {

                $collection = $this->client->selectCollection("myfiles", "fs.files");
                $filter = [];
                $options = ['limit' => $limit];
                return $collection->find($filter, $options);
        }
}

$id = isset($_GET['id']) ? urldecode($_GET['id']) : null;


$gfs = new GridFS();

if(isset($id)) {
        $gfs->Image($id);
}

?>

<html>
<head><title>ImageFiles</title></head>
<body>

<?php

$files = $gfs->FindFiles(10);

foreach ($files as $file) {
        $id = $file["_id"];
        echo "<h2>ID: $id</h2>\r\n";
        echo "<img src='?id=$id' border='0' alt='' /><br /> \r\n";
}
?>

</body>
</html>
```

My solution is, to hide the warning...

```
public function close()
{
    if (is_resource($this->buffer)) {
        @fclose($this->buffer);
    }
}
```

**Comment by jmikola** [comment]
Can you share a PHP script that reproduces this warning?

Based on the code, it's not clear to me how this warning would be triggered unless `ReadableStream::close()` was invoked multiple times, which itself sounds like a bug in our StreamWrapper implementation.

**Comment by IT-Experte** [comment]
I have inserted my example php code. You need some image files into mongo database "myfiles". I think, Close(); was invoked multiple times (for each image).

**Comment by jmikola** [comment]
I was able to test this a bit and it appears to be specific to PHP 7.x. Our StreamWrapper's `close()` method is still only called once, but the in-memory stream we use for buffering appears to be closed before the GridFS stream that encapsulates it. When the `exit;` statement is moved out of the `Image()` method, the script completes without any warning:

```
if(isset($id)) {
    $gfs->Image($id);
    exit; // relocated from GridFS::Image()
}
```

This makes sense, as our stream goes out of scope when `Image()` returns and is closed naturally.

I'm currently in the midst of refactoring the StreamWrappers to not use streams for buffer (see: https://github.com/mongodb/mongo-php-library/pull/304), so this will be resolved. In the meantime, feel free to use the above work-around.

**Comment by IT-Experte** [comment]
Many thanks for your help. I've looked your refactoring and your benchmarks. It is surprising that a string buffer is faster than writing it directly into RAM.

**Comment by jmikola** [comment]
> It is surprising that a string buffer is faster than writing it directly into RAM.

Strings are in RAM as well. I was a bit surprised to find that the `php://memory` stream was nearly 4x slower, but I suppose that may be overhead of PHP's APIs (e.g. `fwrite()`). There is definitely a use case for `php://memory`, as in our tests when we want to quickly feed strings into GridFS without creating a file on disk; however, there wasn't a huge need for it in our StreamWrappers.

**Comment by jmikola** [comment]
This should be fixed by #304 and will be resolved in the next 1.1.x release.

**Comment by jmikola** [comment]
Correction: #304 pertained only to WritableStream refactoring. This was actually resolved by ReadableStream refactoring in https://github.com/mongodb/mongo-php-library/pull/323 ([PHPLIB-247](https://jira.mongodb.org/browse/PHPLIB-247)).
